### PR TITLE
bpo-41911: Update docs for various expressions

### PR DESCRIPTION
**Pull request by akulakov**
https://bugs.python.org/issue41911

### CODE REVIEW

File: Doc/reference/expressions.rst [hunk 1]
```diff
@@ -1174,22 +1175,25 @@ All unary arithmetic and bitwise operations have the same priority:
...
 
 .. index::
    single: inversion
    operator: ~ (tilde)
 
 The unary ``~`` (invert) operator yields the bitwise inversion of its integer
 argument.  The bitwise inversion of ``x`` is defined as ``-(x+1)``.  It only
-applies to integral numbers.
+applies to integral numbers or to custom objects that override the
+:meth:`__negate__` special method.
```
Comment (ambv) [inline Doc/reference/expressions.rst]: What is `__negate__`? You mean `__invert__`.

File: Doc/reference/expressions.rst [hunk 2]
```diff
@@ -1221,9 +1225,11 @@ operators and one for additive operators:
...
 The ``*`` (multiplication) operator yields the product of its arguments.  The
 arguments must either both be numbers, or one argument must be an integer and
-the other must be a sequence. In the former case, the numbers are converted to a
-common type and then multiplied together.  In the latter case, sequence
-repetition is performed; a negative repetition factor yields an empty sequence.
+the other must be a sequence, or at least one of them must be a custom object
+implementing the respective :meth:`__mul__` or :meth:`__rmul__` method. In the
+first case, the numbers are converted to a common type and then multiplied
+together.  In the second case, sequence repetition is performed; a negative
+repetition factor yields an empty sequence.
```
Comment (ambv) [inline Doc/reference/expressions.rst]: I see what you're trying to say here but this became less readable now. Former/latter is preferred over first/second. More importantly, when you now mention custom objects with `__mul__` or `__rmul__`, the sentence after seems like it's talking about *those methods* and not the either-or.

Instead of this, I would do what you do in other places: add a single-sentence paragraph saying that the operation can be customized.

File: Doc/reference/expressions.rst [hunk 3]
```diff
@@ -1255,10 +1263,10 @@ The ``%`` (modulo) operator yields the remainder from the division of the first
...
 type.  A zero right argument raises the :exc:`ZeroDivisionError` exception.  The
 arguments may be floating point numbers, e.g., ``3.14%0.7`` equals ``0.34``
-(since ``3.14`` equals ``4*0.7 + 0.34``.)  The modulo operator always yields a
-result with the same sign as its second operand (or zero); the absolute value of
-the result is strictly smaller than the absolute value of the second operand
-[#]_.
+(since ``3.14`` equals ``4*0.7 + 0.34``.)  The modulo operator, unless overridden
+with the special :meth:`__mod__` method, always yields a result with the same sign
+as its second operand (or zero); the absolute value of the result is strictly
+smaller than the absolute value of the second operand [#]_.
```
Comment (ambv) [inline Doc/reference/expressions.rst]: Again, I feel like mixing basic information with the ability to override with a special method, makes the entire thing less readable.

File: Doc/reference/expressions.rst [hunk 4]
```diff
@@ -1279,10 +1287,11 @@ point number using the :func:`abs` function if appropriate.
...
 
-The ``+`` (addition) operator yields the sum of its arguments.  The arguments
-must either both be numbers or both be sequences of the same type.  In the
-former case, the numbers are converted to a common type and then added together.
-In the latter case, the sequences are concatenated.
+The ``+`` (addition) operator yields the sum of its arguments.  The arguments must
+either both be numbers or both be sequences of the same type, or either of them
+must be a custom object that overrides :meth:`add` or :meth:`radd` respectively.
+In the first case, the numbers are converted to a common type and then added
+together.  In the second case, the sequences are concatenated.
```
Comment (ambv) [inline Doc/reference/expressions.rst]: Ditto. Plus you mean dunder-add and dunder-radd.

File: Doc/reference/expressions.rst [hunk 5]
```diff
@@ -1308,8 +1319,10 @@ The shifting operations have lower priority than the arithmetic operations:
 .. productionlist:: python-grammar
    shift_expr: `a_expr` | `shift_expr` ("<<" | ">>") `a_expr`
 
-These operators accept integers as arguments.  They shift the first argument to
-the left or right by the number of bits given by the second argument.
+These operators accept integers as arguments or custom objects that override
+:meth:`__lshift__` and :meth:`__rshift__` special methods.  In the former case they
+shift the first argument to the left or right by the number of bits given by the
+second argument.
```
Comment (ambv) [inline Doc/reference/expressions.rst]: Ditto, don't mix advanced information with the basic information. It gets confusing.

File: Doc/reference/expressions.rst [hunk 6]
```diff
@@ -1335,24 +1348,27 @@ Each of the three bitwise operations has a different priority level:
    pair: bitwise; and
    operator: & (ampersand)
 
-The ``&`` operator yields the bitwise AND of its arguments, which must be
-integers.
+The ``&`` operator yields the bitwise AND of its arguments, which must be integers
+or one of them must be a custom object overriding :meth:`__and__` or
+:meth:`__rand__` special methods.
```
Comment (ambv) [inline Doc/reference/expressions.rst]: This is OK.

File: Doc/reference/expressions.rst [hunk 7]
```diff
@@ -1335,24 +1348,27 @@ Each of the three bitwise operations has a different priority level:
...
 
 .. index::
    pair: bitwise; xor
    pair: exclusive; or
    operator: ^ (caret)
 
 The ``^`` operator yields the bitwise XOR (exclusive OR) of its arguments, which
-must be integers.
+must be integers or one of them must be a custom object overriding :meth:`__xor__` or
+:meth:`__rxor__` special methods.
```
Comment (ambv) [inline Doc/reference/expressions.rst:1371]: OK.

File: Doc/reference/expressions.rst [hunk 8]
```diff
@@ -1335,24 +1348,27 @@ Each of the three bitwise operations has a different priority level:
...
 
 .. index::
    pair: bitwise; or
    pair: inclusive; or
    operator: | (vertical bar)
 
 The ``|`` operator yields the bitwise (inclusive) OR of its arguments, which
-must be integers.
+must be integers or one of them must be a custom object overriding :meth:`__or__` or
+:meth:`__ror__` special methods.
```
Comment (ambv) [inline Doc/reference/expressions.rst:1380]: OK.

File: Doc/reference/expressions.rst [hunk 9]
```diff
@@ -1380,7 +1396,9 @@ in mathematics:
    comp_operator: "<" | ">" | "==" | ">=" | "<=" | "!="
                 : | "is" ["not"] | ["not"] "in"
 
-Comparisons yield boolean values: ``True`` or ``False``.
+Comparisons normally yield boolean values: ``True`` or ``False``, but
+:dfn:`rich comparison methods` not in a context of :dfn:`Boolean operations` may
+yield anything.
```
Comment (ambv) [inline Doc/reference/expressions.rst]:
```suggestion
Comparisons yield boolean values: ``True`` or ``False``. Custom
:dfn:`rich comparison methods` may return non-boolean values. In this case
Python will call :func:`bool` on such value to convert it to boolean as needed.
```

File: Doc/reference/expressions.rst (hunk 1 above)
Comment (akulakov) [reply Doc/reference/expressions.rst]: You are right, I have no idea what I was thinking :)

File: Doc/reference/expressions.rst [hunk 10]
```diff
@@ -1380,7 +1405,9 @@ in mathematics:
    comp_operator: "<" | ">" | "==" | ">=" | "<=" | "!="
                 : | "is" ["not"] | ["not"] "in"
 
-Comparisons yield boolean values: ``True`` or ``False``.
+Comparisons yield boolean values: ``True`` or ``False``. Custom
+:dfn:`rich comparison methods` may return non-boolean values. In this case
+Python will call :func:`bool` on such value to convert it to boolean as needed.
```
Comment (akulakov) [inline Doc/reference/expressions.rst]: @ambv on a 2nd thought, the problem is this phrasing can be misleading:

[ins] In [7]: class A:
         ...:     def __lt__(s, o):
         ...:         return 'ay'
         ...:
         ...:

[ins] In [8]: A() < 5
Out[8]: 'ay'

Current text can be read to imply bool() would be called before this value is printed or used somewhere, it's not obvious IMO that 'as needed' refers only to `if` constructs.

Comment (ambv) [reply Doc/reference/expressions.rst]: "as needed", e.g. when a boolean context is expected. But that gets too wordy to spell out and repeats the word "boolean" like three times.

File: Doc/reference/expressions.rst [hunk 11]
```diff
@@ -1220,10 +1226,13 @@ operators and one for additive operators:
...
 
 The ``*`` (multiplication) operator yields the product of its arguments.  The
-arguments must either both be numbers, or one argument must be an integer and
-the other must be a sequence. In the former case, the numbers are converted to a
-common type and then multiplied together.  In the latter case, sequence
-repetition is performed; a negative repetition factor yields an empty sequence.
+arguments must either both be numbers, or one argument must be an integer and the
+other must be a sequence.  In the former case, the numbers are converted to a
+common type and then multiplied together.  In the latter case, sequence repetition
+is performed; a negative repetition factor yields an empty sequence.
```
Comment (ambv) [inline Doc/reference/expressions.rst]: Ideally there would be no diff for places in text you didn't end up changing.

File: Doc/reference/expressions.rst [hunk 12]
```diff
@@ -1308,8 +1326,12 @@ The shifting operations have lower priority than the arithmetic operations:
 .. productionlist:: python-grammar
    shift_expr: `a_expr` | `shift_expr` ("<<" | ">>") `a_expr`
 
-These operators accept integers as arguments.  They shift the first argument to
-the left or right by the number of bits given by the second argument.
+These operators accept integers as arguments.  In the former case they shift the
+first argument to the left or right by the number of bits given by the second
+argument.
```
Comment (ambv) [inline Doc/reference/expressions.rst]: Leaving this diff made it easy to miss that your text still says "In the former case" here.

File: Doc/reference/expressions.rst (hunk 10 above)
Comment (akulakov) [reply Doc/reference/expressions.rst]: What do you think about "Python will call :func:`bool` on such value in boolean contexts." ?

Comment (ambv) [reply Doc/reference/expressions.rst]: Sounds good, I like it.

**Comment by akulakov** [comment]
@ambv Good point about minimizing diffs!; I fixed them manually but then a minute later realized i could have done it much easier by adjusting 'textwidth' in the editor, - I'll remember to do that next time.

Updated per comments, PTAL..

**Comment by akulakov** [comment]
@ambv thanks for review :)
